backend/pkg/tls/ca.go

```diff
@@ -0,0 +1,134 @@
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+
+package tls
+
+import (
+	"context"
+	"crypto/sha256"
+	libtls "crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// MaybeWithDynamicClientCA returns a function that configures a CA on the tls
+// Config reading it from a file and updating whenever its content changes.
+// It is intended to be used on a client configuration to set the server CA it
+// expects to talk to.
+//
+// Asynchronous reloading is triggered when a new request is served, not before
+// 5 minutes from last reload.
+func MaybeWithDynamicClientCA(
+	quitCtx context.Context,
+	caPath string,
+	hostname string,
+	refreshInterval time.Duration,
+	logger *zap.Logger,
+) func(*libtls.Config) error {
+	return func(cfg *libtls.Config) error {
+		if caPath == "" {
+			return nil
+		}
+
+		r, err := newCAReloader(quitCtx, cfg, caPath, refreshInterval, logger)
+		if err != nil {
+			return err
+		}
+		// There's no direct client-side equivalent to GetConfigForClient;
+		// instead, we skip the standard server certificate validation and
+		// override VerifyConnection instead.
+		cfg.InsecureSkipVerify = true
+		cfg.VerifyConnection = func(s libtls.ConnectionState) error {
+			root := r.tc.Load()
+			pool := root.ClientCAs
+			leaf := s.PeerCertificates[0]
+			_, err := leaf.Verify(x509.VerifyOptions{
+				DNSName: hostname,
+				Roots:   pool,
+			})
+			return err
+		}
+		return nil
+	}
+}
+
+type caReloader struct {
+	tc     atomic.Pointer[libtls.Config]
+	digest [sha256.Size]byte
+
+	caPath string
+	logger *zap.Logger
+}
+
+func newCAReloader(quitCtx context.Context, initCfg *libtls.Config, caPath string, refreshInterval time.Duration,
+	logger *zap.Logger,
+) (*caReloader, error) {
+	r := caReloader{
+		caPath: caPath,
+		logger: logger,
+	}
+	r.tc.Store(initCfg)
+	if err := r.load(); err != nil {
+		return nil, err
+	}
+	go r.reloader(quitCtx, refreshInterval)
+	return &r, nil
+}
+
+func (r *caReloader) reloader(quitCtx context.Context, refreshInterval time.Duration) {
+	ticker := time.NewTicker(refreshInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-quitCtx.Done():
+			return
+		case <-ticker.C:
+			if err := r.load(); err != nil {
+				r.logger.Error("error while reloading the CA", zap.Any("err", err))
+			}
+		}
+	}
+}
+
+func (r *caReloader) load() error {
+	ca, err := os.ReadFile(r.caPath)
+	if err != nil {
+		return fmt.Errorf("cannot read CA file %q: %w", r.caPath, err)
+	}
+	digest := sha256.Sum256(ca)
+	if digest == r.digest {
+		r.logger.Debug("CA cert did not change", zap.String("path", r.caPath))
+		return nil
+	}
+	cfg := r.tc.Load().Clone()
+	if cfg.ClientCAs == nil {
+		cfg.ClientCAs = x509.NewCertPool()
+	}
+	if !cfg.ClientCAs.AppendCertsFromPEM(ca) {
```

Comment:
I understand this fn add the new ca to the existing ones, do we remove the old ones at some point?

Reply:
We don't, this is how Serverless does it and it makes sense imho. Certificates will be rotated by kubernetes because they will be expired soon, if the reason for rotation is a certificate breach I would expect the pod to be restarted.